tests/utilities/test_collections.py

```diff
@@ -400,6 +407,44 @@ def test_visit_collection_max_depth(self, inp, depth, expected):
         )
         assert result == expected
 
+    def test_visit_collection_context(self):
+        foo = [1, 2, [3, 4], [5, [6, 7]], 8, 9]
+
+        def visit(expr, context):
+            if isinstance(expr, list):
+                context["depth"] += 1
+                return expr
+            else:
+                return expr + context["depth"]
+
+        result = visit_collection(foo, visit, context={"depth": 0}, return_data=True)
+        assert result == [2, 3, [5, 6], [7, [9, 10]], 9, 10]
```

Comment:
nit - a comment here giving a tldr of why we're expecting this, or a short docstring in the visit function would reduce the overhead and context needed to understand the behavior of `visit_collection` from this test.